contracts/external/cw-vesting/src/contract.rs

```diff
@@ -0,0 +1,549 @@
+
+    let mut msgs: Vec<CosmosMsg> = vec![];
+
+    let vesting_payment =
+        VESTING_PAYMENT.update(deps.storage, |mut vp| -> Result<_, ContractError> {
+            // Check if contract status is active
+            if vp.status != VestingPaymentStatus::Active {
+                return Err(ContractError::NotActive);
+            }
+
+            // Set status based on whether any funds are staked
+            if vp.staked_amount == Uint128::zero() {
+                vp.status = VestingPaymentStatus::Canceled;
+            } else {
+                vp.status = VestingPaymentStatus::CanceledAndUnbonding;
+                // Set canceled at time for use when distributing unbonded funds
+                vp.canceled_at_time = Some(env.block.time.seconds());
+            }
+            Ok(vp)
+        })?;
+
+    // In no funds are staked, we distibute contract funds
+    // Check if funds have vested or there are staking rewards
+    let vested_amount = vesting_payment.get_vested_amount_by_seconds(env.block.time.seconds())?;
+
+    // Transfer any vested amount to the original recipient
+    if vested_amount != Uint128::zero() {
+        let transfer_vested_to_recipient_msg = vesting_payment
+            .denom
+            .get_transfer_to_message(&vesting_payment.recipient, vested_amount)?;
+        msgs.push(transfer_vested_to_recipient_msg);
+    }
+
+    // Transfer any remaining unvested amount to the owner
+    let unvested = vesting_payment.amount.checked_sub(vested_amount)?;
+    if unvested != Uint128::zero() {
+        let transfer_unvested_to_owner_msg = vesting_payment
+            .denom
+            .get_transfer_to_message(&info.sender, unvested)?;
+        msgs.push(transfer_unvested_to_owner_msg);
+    }
+
+    // Handle edge case if funds are staked at the time of cancelation
+    if vesting_payment.status == VestingPaymentStatus::CanceledAndUnbonding {
+        let denom = deps.querier.query_bonded_denom()?;
+
+        // Unbond any staked funds
+        let mut undelegate_msgs = STAKED_VESTING_BY_VALIDATOR
+            .range(deps.storage, None, None, Order::Ascending)
+            .flat_map(|svc| -> StdResult<CosmosMsg> {
+                let (validator, amount) = svc?;
+                Ok(CosmosMsg::Staking(StakingMsg::Undelegate {
+                    validator,
+                    amount: Coin {
+                        denom: denom.clone(),
+                        amount,
+                    },
+                }))
+            })
+            .collect::<Vec<CosmosMsg>>();
+
+        msgs.append(&mut undelegate_msgs);
+    }
+
+    Ok(Response::new()
+        .add_attribute("method", "remove_vesting_payment")
+        .add_attribute("owner", info.sender)
+        .add_attribute("removed_time", env.block.time.to_string())
+        .add_messages(msgs))
+}
+
+pub fn execute_distribute(env: Env, deps: DepsMut) -> Result<Response, ContractError> {
+    let vesting_payment = VESTING_PAYMENT.load(deps.storage)?;
```

Comment:
we need a "is sender the recipient" check here too

Reply:
Don't think so. That way this can be automated by Croncat for example.

Reply:
Payment always goes to the recipient so it should be fine.

Reply:
makes sense